iaso/api/org_units.py

```diff
@@ -428,25 +429,20 @@ def partial_update(self, request, pk=None):
             org_unit_type = get_object_or_404(OrgUnitType, id=org_unit_type_id)
             org_unit.org_unit_type = org_unit_type
 
-        if "reference_instance_id" in request.data:
-            reference_instance_id = request.data["reference_instance_id"]
-            if reference_instance_id:
-                instance = Instance.objects.get(pk=reference_instance_id)
-                # Check if the instance has as form the reference_form for the orgUnittype
-                # if the reference_form is the same as the form related to the instance one,
-                # assign the instance to the orgUnit as reference instance
-                if org_unit.org_unit_type.reference_form != instance.form:
+        if "reference_instance_id" in request.data and "reference_instance_action" in request.data:
+            instance = get_object_or_404(Instance, pk=request.data["reference_instance_id"], org_unit=org_unit)
+            if request.data["reference_instance_action"] == Instance.REFERENCE_FLAG_CODE:
+                try:
+                    instance.flag_reference_instance(org_unit)
+                except ValidationError as e:
                     errors.append(
                         {
-                            "errorKey": "reference_form",
-                            "errorMessage": _("Form of subimssion is not allowed on this type of org unit"),
+                            "errorKey": "reference_instances",
+                            "errorMessage": e.messages[0],
                         }
                     )
-                else:
-                    org_unit.reference_instance = instance
-            else:
-                instance = None
-                org_unit.reference_instance = instance
+            elif request.data["reference_instance_action"] == Instance.REFERENCE_UNFLAG_CODE:
+                instance.unflag_reference_instance(org_unit)
```

Comment:
There should probably be a `else` to add an error to the `errors` dict because a wrong action was provided